File: gtci/09-hashmaps/cpp/1-first-non-repeating-character.cpp
#include <cassert>
#include <iostream>
#include <unordered_map>

using namespace std;

class Solution {
   public:
    int firstUniqChar(string s) {
        std::unordered_map<char, int> freqMap;

        for (char ch : s) freqMap[ch] += 1;

        for (int i = 0; i < s.size(); i++) {
            if (freqMap[s[i]] == 1) return i;
        }

        return -1;
    }
};

int main() {
    Solution sol;

    assert(sol.firstUniqChar("leetcode") == 0);
    assert(sol.firstUniqChar("abcab") == 2);
    assert(sol.firstUniqChar("abab") == -1);

    std::cout << "All test cases passed." << std::endl;

    return 0;
}

File: gtci/09-hashmaps/python3/1-first-non-repeating-character.py
class Solution:
    def firstUniqueChar(self, s: str) -> int:
        freq_map = {}

        for idx, char in enumerate(s):
            if char not in freq_map:
                freq_map[char] = 0
            freq_map[char] += 1

        for idx, char in enumerate(s):
            if freq_map[char] == 1:
                return idx

        return -1


if __name__ == "__main__":
    sol = Solution()

    assert sol.firstUniqueChar("leetcode") == 0
    assert sol.firstUniqueChar("abcab") == 2
    assert sol.firstUniqueChar("abab") == -1

    print("All test cases passed.")
